File: 238 Product of Array Except Self/productExceptSelf.py
class Solution:
  def productExceptSelf(self, nums: List[int]) -> List[int]:
    n = len(nums)
    incum = list(nums)    # increasing cum prod; incum[n-1] is not used
    decum = list(nums)    # decreasing cum prod; decum[0] is not used
    for i in range(1, n-1):
      incum[i] = incum[i-1] * nums[i]
      decum[-i-1] = decum[-i] * nums[-i-1]
    result = list(nums)
    result[0] = decum[1]
    result[-1] = incum[-2]
    for i in range(1, n-1):
      result[i] = incum[i-1] * decum[i+1]
    return result

File: 238 Product of Array Except Self/product_except_self.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len(); // `n` is a runtime-determined value
        let mut incum = vec![1; n]; // Dynamic vector to track prefix products
        let mut decum = vec![1; n]; // Dynamic vector to track suffix products
        let mut result = vec![1; n];
        for i in 1..n {
            incum[i] = incum[i-1] * nums[i-1];
        }
        for i in (0..n-1).rev() {
            decum[i] = decum[i+1] * nums[i+1];
        }
        for i in 0..n {
            result[i] = incum[i] * decum[i];
        }
        result
    }
}